photon-core/src/main/java/org/photonvision/common/dataflow/networktables/TimeSyncManager.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (C) Photon Vision.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package org.photonvision.common.dataflow.networktables;
+
+import edu.wpi.first.cscore.CameraServerJNI;
+import edu.wpi.first.networktables.IntegerPublisher;
+import edu.wpi.first.networktables.NetworkTable;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import org.photonvision.common.configuration.NetworkConfig;
+import org.photonvision.common.logging.LogGroup;
+import org.photonvision.common.logging.Logger;
+import org.photonvision.common.util.TimedTaskManager;
+import org.photonvision.jni.PhotonTargetingJniLoader;
+import org.photonvision.jni.TimeSyncClient;
+import org.photonvision.jni.TimeSyncServer;
+
+public class TimeSyncManager {
+    private static final Logger logger = new Logger(TimeSyncManager.class, LogGroup.NetworkTables);
+
+    private TimeSyncClient m_client = null;
+    private TimeSyncServer m_server = null;
+
+    private NetworkTableInstance ntInstance;
+    IntegerPublisher m_offsetPub;
+    IntegerPublisher m_rtt2Pub;
+    IntegerPublisher m_pingsPub;
+    IntegerPublisher m_pongsPub;
+    IntegerPublisher m_lastPongTimePub;
+
+    public TimeSyncManager(NetworkTable kRootTable) {
+        if (!PhotonTargetingJniLoader.isWorking) {
+            logger.error("PhotonTargetingJNI was not loaded! Cannot do time-sync");
```

Comment:
What does the timestamp look like to the roborio if the jni is not working?

Reply:
it'll be in the coprocessor timebase, which is us since coproc boot. So I think basically unusable? I'm lazy, but we should be robust to or loud about this.

Reply:
Maybe we could send a negative time and just swap it for fpga now. Idk how likely is it we hit a case like this on supported hardware.

Reply:
yeah, or a single "time offset valid" bool? and then photonlib side, swap between trusting recieve timesamp and the encoded timestamp? we def should pass back a time sync health flag up to robot code though.

Reply:
Added - you happy with the new guards?

Reply:
I think a check for wether the camera has had a pong would be good and if it hasn't we replace with fpga now and continue along our way with a big old warning